File: leetcode/problems_list/238-ProductOfArrayExceptSelf/solution.py
"""
Given an array nums of n integers where n > 1,  return an array output such that output[i] is equal to the product of all the elements of nums except nums[i].
"""
# Time complexity: O(n)
# Space complexity: O(1)


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1]*len(nums)
        left = 1
        for i in range(len(nums)):
            res[i] = left
            left *= nums[i]
        right = 1
        for i in range(len(nums)-1, -1, -1):
            res[i] *= right
            right *= nums[i]

        return res

File: leetcode/problems_list/238-ProductOfArrayExceptSelf/solution.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
// Time complexity: O(n)
// Space complexity: O(1)
var productExceptSelf = function (nums) {
  const res = [];
  let temp = 1;
  for (let i = 0; i < nums.length; i++) {
    res[i] = temp;
    temp *= nums[i];
  }
  temp = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    res[i] *= temp;
    temp *= nums[i];
  }
  return res;
};
